File: lab3-2.cpp
#include <iostream>
#include <fstream>
#include <cstring>

using namespace std;

const int Stack_Size = 20;

class Stack {
    private:
        char stack[Stack_Size]; int top;
    public:
        Stack() {
                top = -1;
        }
        void push(int val) {
                stack[++top] = val;
        }
        char pop() {
                return stack[top--];
        }
        int isEmpty() {
                return top == -1;
        }
        int isFull() {
                return top == Stack_Size - 1;
        }
        };


int main()
{
        Stack s1;
        ifstream in;
        in.open("lab3-2.txt");
        char buffer[80];
        while(in.getline(buffer, 80)){
            int i = 0;
            // cout << buffer << endl;
            if(strlen(buffer) % 2 == 0){
                for(; i < strlen(buffer)/2; i++)
                    s1.push(buffer[i]);
            }
            else{
                for(; i < strlen(buffer)/2; i++)
                    s1.push(buffer[i]);
                i++;
            }
            while(buffer[i] != '\0'){
                if(buffer[i] != s1.pop()){
                    cout << "error" << endl;
                    break;
                }
                i++;
            }
            if(buffer[i] == '\0'){
                cout << "PALINDROME" << endl;
            }
        }
}

File: lab3-2.py
f = open('lab3-2.txt', 'r')


def push(val):
    global top
    top += 1
    stack[top] = val


def pop():
    global top
    fin = top
    top -= 1
    return stack[fin]


def isEmpty():
    global top
    return top == -1


def isFull():
    global top
    return top == len(stack)


if __name__ == "__main__":
    a = f.read()
    top = -1
    stack = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    print("check stack full: " + str(isFull()))
    print("check stack empty: " + str(isEmpty()))
    print()
    for i in range(len(a.split())):
        data = a.split()[i]
        j = 0
        if (len(data) % 2 == 0):
            while j < (len(data) // 2):
                push(data[j])
                j += 1
        else:
            while j < (len(data) // 2):
                push(data[j])
                j += 1
            j += 1
        while j < len(data):
            if (pop() == data[j]):
                j += 1
            else:
                print(data + " Error")
                break
        if j == len(data):
            print(data + " PALINDROME")
